### PR TITLE
Cleanup code for constraint optimization

### PR DESCRIPTION
**Pull request by harshjk**
This PR resolves issue #17869

Test plan - Unit Tests

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java [hunk 1]
```diff
@@ -388,7 +382,7 @@ LogicalPropertiesImpl build()
             requireNonNull(sourceProperties, "sourceProperties is null");
             requireNonNull(keyVariables, "keyVariables is null");
             requireNonNull(outputVariables, "outputVariables is null");
-            checkArgument(!keyVariables.isEmpty(), "keyVariables is empty");
+            Preconditions.checkArgument(!keyVariables.isEmpty(), "keyVariables is empty");
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java:385]: `import static com.google.common.base.Preconditions.checkArgument;` and remove `Preconditions.` from `Preconditions.checkArgument`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java [hunk 2]
```diff
@@ -431,7 +425,7 @@ LogicalPropertiesImpl build()
             requireNonNull(keyVariables, "keyVariables is null");
             requireNonNull(outputVariables, "outputVariables is null");
             requireNonNull(limit, "limit is null");
-            checkArgument(!keyVariables.isEmpty(), "keyVariables is empty");
+            Preconditions.checkArgument(!keyVariables.isEmpty(), "keyVariables is empty");
```
Comment (yingsu00) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java:428]: ditto

**Review by yingsu00 — COMMENTED** [review]
@harshjk Thank you for submitting the PR! The changes look good, but the following is missing:
1. EquivalenceClassProperty.java, move line 231-232 to the caller
2. EquivalenceClassProperty.java, remove unnecessary toString() on Expressions in the toString() method
3. KeyProperty.java Remove line 113 `requireNonNull(newKey, "newKey is null");`
4. Replace the builder classes in LogicalPropertiesImpl to static methods
5. RemoveRedundantAggregateDistinct.java, simplify apply() by replacing the streaming API to for loop. The current implementation is hard to read. Also, @rongrong mentioned to check match once here

I wonder if you have time to resolve them? It will be great if you finish them as well, but if not we will deal with it later.

Also, could you please squash your two commits into one?

**Comment by yingsu00** [comment]
One more thing, we don't typically refer to issue number in the commit title. You can mention that in the commit message.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java (hunk 1 above)
Comment (harshjk) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java]: Resolved in latest commit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java (hunk 2 above)
Comment (harshjk) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/properties/LogicalPropertiesImpl.java]: Resolved in the latest commit.

**Comment by harshjk** [comment]
> One more thing, we don't typically refer to issue number in the commit title. You can mention that in the commit message.

Thank you for your review, and suggestions, and for updating the PR title. 

I would love to address your additional suggestion, but I cannot at that time.

**Comment by harshjk** [comment]
@yingsu00 Can you please review and approve the pull request?

**Comment by yingsu00** [comment]
> 4\. LogicalPropertiesImpl

@harshjk Thank you for updating the PR. I see that you addressed the Preconditions comment, but didn't try the ones in the list. Are you still able to address them? 

Also would you please squash the commits into one single commit? Thank you!

**Comment by yingsu00** [comment]
@harshjk Will you be able to squash the commits into one? It looks like you're not working on this any more. I'll take over your PR if I don't hear back from you this Friday. 
cc @rschlussel

**Comment by rschlussel** [comment]
Another thing that should be cleaned up for constraint optimization is to make all the property fields immutable, and instead create them in their final state.  There isn’t really a good reason for them to be constructed and then updated. Most actually start with an empty property and then update once to some source properties.  There are a few that combine property classes, but it would be better to have a way to combine that constructs a new combined instance rather than the mutable objects we currently have.  They never need to be mutated after they are fully constructed by the logical properties builders, so we should make them immutable.  

I started this before I realize there is other refactoring work in progress.  You're welcome to base off of whatever that if it's useful.  KeyProperty and MaxCardProperty are mostly done.  EquivalenceClassProperty still needs work. https://github.com/rschlussel/presto/commit/b7d5228efc41e010e0d37edc045a7d7c9928b9bc

**Comment by yingsu00** [comment]
Closing in favor of https://github.com/prestodb/presto/pull/18192

**Pull request closed without merging**